README.md
# IMR: Extended

IMR: Extended is **LIVE** (In Alpha, prone to bugs, NaNs, save loss, etc... Play at your own risk)
Introduces (At the moment):

\*Whole mod:

-Harder to gain currencies:

Rage Power (1e27 g of mass from 1e20 g of mass)
Atoms (1e175 g of Black Hole from 1e156 g of Black Hole)

\*Main tab:

-New ranks, tiers, tetrs and pents

\*Supernova neutron star tree main tab:

-New upgrades

\*Atom mass dilation tab:

-Some effects changed

\*Quantum entropy tab:

-New upgrades

---

For the (FAR) future:

-More mass upgrades

-More ranks and prestiges rewards

-New types of ranks post Enne and new prestiges post Renown

-More upgrades post big rip

-More challenges post c16

-More elements and element tiers

-More mass dilation and break dilation upgrades

-New mass dilation called Dimensional Dilation (Idea on how it will be, will be discussed another time)

-New currency called Anti-Quarks

-More neutron tree upgrades and tabs

-More bosons

-More fermions

-Even more radiation

-More quantum milestones

-More primordium particles

-More entropy upgrades

-More dark effects

-More dark runs upgrades and glyphs

-The final shard is not the end!

-More corruption upgrades

-New mechanics & layers!

And much much more in the future!

---

Play the Alpha now in https://officialzygorg.github.io/incremental-mass-rewritten/


changelog
0.1.8 6/25/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Added:
-More prestiges,honors.
Modified:
-Big Rip upgrade costs.
-Nerfed Quark gain while big rip.
-Nerfed Break Dilation relativistic energy gain.
-Some scalings.
-Changed qc2 in the supernova tree
-QC challenges nerf (they're easier)
Fixed:
-Break Dilation energy gain NaNing if you exit and enter Big Rip while you're breaking dilation.
-Quantum challenge UI no telling you can go past 10 in a challenge.
-Prestige 4 NaNing atom.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the link in the options tab.
You can check the version updates also on the options tab.

0.1.7-F1 6/25/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Fixed:
-Going beyond 19 on the last QC modifier breaks the game.
-Tier 30000 description.
Modified:
-Last QC modifier nerf.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.7 6/24/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Modified:
-Nerfed Quantum Challenges scalings
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.6 6/24/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Added: More ranks,tiers,tetrs,pents
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.5-F3 6/24/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Fixed:
-Rank 1,2 rewards we're multiplying by 0 the mass upgrade 1,2 effects if you had less than 10 bought.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.5-F2 6/24/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Modified:
-Star gain in CH16 is now ^0.1
-Changed description of CH16
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.5-F1 6/24/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Fixed:
-Hawking radiation evaporation being weak.
-qu_qol7 not working.
-Stars going nuts on CH16
-Free FFS on CH16
-Tier autobuyer now goes FAST.
Modified:
-Softcapped rank 3000 effect just in case it makes last star go nuts.
-Updated break_eternity.js to the latest version as today and fixed Decimal.clone being absent in it.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.5 6/23/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Added:
-More pents.
Modified:
-Rank 1 and 2 effects now affect their correspondent upgrade effect instead of adding more free mass upgrades. (Also, due to this, rank 9 is nerfed.)
Removed: some ranks due to the change of rank 1 and 2 being modified.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.4-F1 6/23/2023 IMR:Extended fix
https://officialzygorg.github.io/imr-z/
Fixed:
-Last star (blue star) going nuts with rank 1300 effect

0.1.4 6/23/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Modified:
-Some ranks (balance)
Fixed?:
-Game gets NaNed at primordium/1 manual quantize/20 quantizes
Added:
-More ranks,tiers
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.3 6/22/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Modified:
-Quantum Foam requierements from e4e14 to e1e12
-Atom requirements from 1.5e175 to 1.5e160 because people couldnt reach it.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.2 6/22/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/
Fixed:
-Rank 25 not working
Modified:
-Rank 1300 (Stronger)
Added:
-More ranks,tiers,tetrs,pents
Mod endgame: rank 24000, tier 1000, tetr 75, pent 6, qa4 in supernova tree.
Any inflations, bugs or NaNs, pin me on discord or open an issue in the github.

0.1.1 6/22/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/

Modified:
-Challenge 10 1st goal from 1e30056 to 1e17556

Any inflations, bugs or NaNs, pin me on discord or open an issue in https://github.com/OfficialZygorg/imr-z/issues

0.1.0-F2 6/22/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/

-Fixes:
Rank 1300 NaNing the crap outta the game.

0.1.0-F1 6/22/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/

-Fixes:
Rank 125 NaNed dark matter gain.

Endgame: Unknown at the moment

Any inflations, bugs or NaNs, pin me on discord or open an issue in https://github.com/OfficialZygorg/imr-z/issues

0.1.0 6/22/2023 IMR:Extended update
https://officialzygorg.github.io/imr-z/

-Ported the mod to the infinity update (look at 20.06.23.txt)

Endgame: Unknown at the moment

Any inflations, bugs or NaNs, pin me on discord or open an issue in https://github.com/OfficialZygorg/imr-z/issues


js/scaling.js
const SCALE_START = {
  super: {
    rank: E(50),
    tier: E(10),
    tetr: E(7),
    pent: E(15),
    hex: E(10),
    massUpg: E(100),
    tickspeed: E(100),
    bh_condenser: E(100),
    gamma_ray: E(100),
    supernova: E(15),
    fTier: E(10),
    cosmic_str: E(15),
    prestige0: E(15),
    prestige1: E(7),
    prestige2: E(30),
    prestige3: E(8),
    massUpg4: E(50),
    FSS: E(5),
    pe: E(25),
    inf_theorem: E(10),
  },
  hyper: {
    rank: E(120),
    tier: E(200),
    tetr: E(60),
    pent: E(60),
    hex: E(60),
    massUpg: E(500),
    tickspeed: E(250),
    bh_condenser: E(300),
    gamma_ray: E(300),
    supernova: E(35),
    fTier: E(50),
    cosmic_str: E(90),
    prestige0: E(80),
    prestige1: E(60),
    prestige2: E(60),
    FSS: E(32),
  },
  ultra: {
    rank: E(600),
    tier: E(1e5),
    tetr: E(150),
    pent: E(1200),
    hex: E(2e12),
    massUpg: E(1e11),
    tickspeed: E(700),
    bh_condenser: E(750),
    gamma_ray: E(800),
    supernova: E(60),
    fTier: E(100),
    prestige0: E(320),
    prestige1: E(200),
  },
  meta: {
    rank: E(1e4),
    tier: E(1e9),
    tetr: E(1e10),
    tickspeed: E(5e4),
    bh_condenser: E(1e7),
    gamma_ray: E(1e6),
    supernova: E(100),
    fTier: E(1.25e4),
    prestige0: E(3500),
  },
  exotic: {
    rank: E(1e16),
    tier: E(1e25),
    supernova: E(2e5),
  },
  supercritical: {
    rank: E(1e37),
    supernova: E(1e7),
  },
};

const SCALE_POWER = {
  super: {
    rank: 1.5,
    tier: 1.5,
    tetr: 2,
    pent: 2,
    hex: 2.5,
    massUpg: 2.5,
    tickspeed: 2,
    bh_condenser: 2,
    gamma_ray: 2,
    supernova: 3,
    fTier: 2.5,
    cosmic_str: 2,
    prestige0: 1.5,
    prestige1: 1.5,
    prestige2: 2,
    prestige3: 2,
    massUpg4: 3,
    FSS: 2,
    pe: 2,
    inf_theorem: 2,
  },
  hyper: {
    rank: 2.5,
    tier: 2.5,
    tetr: 3,
    pent: 3,
    hex: 3.5,
    massUpg: 5,
    tickspeed: 4,
    bh_condenser: 2,
    gamma_ray: 4,
    supernova: 3,
    fTier: 4,
    cosmic_str: 4,
    prestige0: 2,
    prestige1: 2,
    prestige2: 3,
    FSS: 3,
  },
  ultra: {
    rank: 4,
    tier: 4,
    tetr: 6,
    pent: 6,
    hex: 8,
    massUpg: 10,
    tickspeed: 7,
    bh_condenser: 4,
    gamma_ray: 6,
    supernova: 5,
    fTier: 6,
    prestige0: 3,
    prestige1: 3,
  },
  meta: {
    rank: 1.0025,
    tier: 1.0000001,
    tetr: 1.00000001,
    tickspeed: 1.001,
    bh_condenser: 1.001,
    gamma_ray: 1.001,
    supernova: 1.025,
    fTier: 1.001,
    prestige0: 1.0025,
  },
  exotic: {
    rank: 15,
    tier: 20,
    supernova: 20,
  },
  supercritical: {
    rank: 50,
    supernova: 75,
  },
};

const SCALE_FP = {
  tickspeed() {
    return [1, 1, 1, tmp.tickspeedFP];
  },
};

const QCM8_SCALES = ["rank", "tier", "tetr", "pent", "hex", "massUpg", "tickspeed", "bh_condenser", "gamma_ray", "supernova", "fTier"];
const PreQ_SCALES = ["rank", "tier", "tetr", "massUpg", "tickspeed", "bh_condenser", "gamma_ray"];
const SCALE_TYPE = ["super", "hyper", "ultra", "meta", "exotic", "supercritical"]; // super, hyper, ultra, meta, exotic
const FULL_SCALE_NAME = ["Super", "Hyper", "Ultra", "Meta", "Exotic", "Supercritical"];

const SCALING_RES = {
  rank(x = 0) {
    return player.ranks.rank;
  },
  tier(x = 0) {
    return player.ranks.tier;
  },
  tetr(x = 0) {
    return player.ranks.tetr;
  },
  pent(x = 0) {
    return player.ranks.pent;
  },
  hex(x = 0) {
    return player.ranks.hex;
  },
  tickspeed(x = 0) {
    return player.tickspeed;
  },
  massUpg(x = 1) {
    return E(player.massUpg[x] || 0);
  },
  bh_condenser(x = 0) {
    return player.bh.condenser;
  },
  gamma_ray(x = 0) {
    return player.atom.gamma_ray;
  },
  supernova(x = 0) {
    return player.supernova.times;
  },
  fTier(x = 0, y = 0) {
    return player.supernova.fermions.tiers[x][y];
  },
  cosmic_str(x = 0) {
    return player.qu.cosmic_str;
  },
  prestige0() {
    return player.prestiges[0];
  },
  prestige1() {
    return player.prestiges[1];
  },
  prestige2() {
    return player.prestiges[2];
  },
  prestige3() {
    return player.prestiges[3];
  },
  massUpg4() {
    return E(player.massUpg[4] || 0);
  },
  FSS() {
    return player.dark.matters.final;
  },
  pe() {
    return player.inf.pe;
  },
  inf_theorem() {
    return player.inf.theorem;
  },
};

const NAME_FROM_RES = {
  rank: "Rank",
  tier: "Tier",
  tetr: "Tetr",
  pent: "Pent",
  hex: "Hex",
  massUpg: "Mass Upgrades 1-3",
  tickspeed: "Tickspeed",
  bh_condenser: "Black Hole Condenser",
  gamma_ray: "Cosmic Ray",
  supernova: "Supernova",
  fTier: "Fermion Tier",
  cosmic_str: "Cosmic String",
  prestige0: "Prestige Level",
  prestige1: "Honor",
  prestige2: "Glory",
  prestige3: "Renown",
  massUpg4: "Overpower",
  FSS: "Final Star Shard",
  pe: "Parallel Extruder",
  inf_theorem: "Infinity Theorem",
};

function updateScalingHTML() {
  let s = SCALE_TYPE[player.scaling_ch];
  // tmp.el.scaling_name.setTxt(FULL_SCALE_NAME[player.scaling_ch])
  if (!tmp.scaling) return;
  for (let x = 0; x < SCALE_TYPE.length; x++) {
    tmp.el["scaling_div_" + x].setDisplay(player.scaling_ch == x);
    if (player.scaling_ch == x) {
      let key = Object.keys(SCALE_START[SCALE_TYPE[x]]);
      for (let y = 0; y < key.length; y++) {
        let have = tmp.scaling[SCALE_TYPE[x]].includes(key[y]);
        tmp.el["scaling_" + x + "_" + key[y] + "_div"].setDisplay(have);
        if (have) {
          tmp.el["scaling_" + x + "_" + key[y] + "_power"].setTxt(format(tmp.scaling_power[SCALE_TYPE[x]][key[y]].mul(100)) + "%");
          tmp.el["scaling_" + x + "_" + key[y] + "_start"].setTxt(format(tmp.scaling_start[SCALE_TYPE[x]][key[y]], 0));
        }
      }
    }
  }
}

function updateScalingTemp() {
  for (let x = 0; x < SCALE_TYPE.length; x++) {
    let st = SCALE_TYPE[x];

    tmp.scaling[st] = [];
    tmp.no_scalings[st] = [];

    let sp = tmp.scaling_power[st],
      ss = tmp.scaling_start[st],
      ns = tmp.no_scalings[st];
    let key = Object.keys(SCALE_START[st]);

    for (let y = 0; y < key.length; y++) {
      let sn = key[y];

      sp[sn] = getScalingPower(x, sn);
      ss[sn] = getScalingStart(x, sn);
      if (noScalings(x, sn)) ns.push(sn);
      else {
        if (sn == "massUpg")
          for (let i = 0; i < UPGS.mass.cols; i++) {
            if (scalingActive(sn, SCALING_RES[sn](i), st)) {
              tmp.scaling[st].push(sn);
              break;
            }
          }
        else if (sn == "fTier")
          for (let i = 0; i < 2; i++)
            for (let j = 0; j < 6; j++) {
              if (scalingActive(sn, SCALING_RES[sn](i, j), st)) {
                tmp.scaling[st].push(sn);
                break;
              }
            }
        else if (scalingActive(sn, SCALING_RES[sn](), st)) tmp.scaling[st].push(sn);
      }
    }
  }
  let sqc8 = [];
  if (!CHALS.inChal(14) && !inDarkRun() && !tmp.c16active && !CHALS.inChal(15)) {
    if (player.mainUpg.br.includes(2)) sqc8.push("massUpg", "rank", "tier", "tetr", "pent", "hex");
    if (player.md.break.active) sqc8.push("bh_condenser", "gamma_ray");
  }
  tmp.scaling_qc8 = sqc8;
}

function scalingActive(name, amt, type) {
  if (tmp.no_scalings[type].includes(name) || SCALE_START[type][name] === undefined) return false;
  return Decimal.gte(amt, tmp.scaling_start[type][name]);
}

function scaleStart(type, name) {
  return tmp.scaling_start[type][name] || SCALE_START[type][name];
}

function getScalingName(name, x = 0, y = 0) {
  if (!NAME_FROM_RES[name]) return;

  let cap = Object.keys(SCALE_START).length;
  let current = "";
  let amt = SCALING_RES[name](x, y);
  for (let n = cap - 1; n >= 0; n--) {
    if (scalingActive(name, amt, Object.keys(SCALE_START)[n])) return capitalFirst(Object.keys(SCALE_START)[n]) + (n == 3 ? "-" : " ");
  }
  return current;
}

function getScalingStart(type, name) {
  let c16 = tmp.c16active;

  let start = SCALE_START[SCALE_TYPE[type]][name];

  if (type == 0) {
    if (name == "rank") {
      if (CHALS.inChal(1) || CHALS.inChal(10)) return E(25);
      start = start.add(tmp.chal ? tmp.chal.eff[1].rank : 0);
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "tier") {
      if (player.mainUpg.atom.includes(5)) start = start.add(10);
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "tetr") {
      if (player.ranks.tier.gte(100)) start = start.add(5);
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "pent") {
      if (hasElement(184)) start = start.mul(elemEffect(184));
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "hex") {
      if (hasElement(184)) start = start.mul(elemEffect(184));
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "massUpg") {
      if (CHALS.inChal(1) || CHALS.inChal(10)) return E(25);
      if (player.mainUpg.bh.includes(3)) start = start.add(tmp.upgs ? (tmp.upgs.main ? tmp.upgs.main[2][3].effect : 0) : 0);
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "tickspeed") {
      if (CHALS.inChal(1) || CHALS.inChal(10)) return E(50);
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "prestige0") {
      if (player.md.break.upgs[9].gte(1)) start = start.add(10);
      if (hasElement(175)) start = start.add(30);
      if (hasElement(194)) start = start.mul(2);
      if (player.ranks.tier.gte(150)) start = start.mul(RANKS.effect.tier[150]());
    } else if (name == "FSS") {
      if (hasBeyondRank(3, 2)) start = start.add(1);
      if (hasBeyondRank(5, 2)) start = start.add(beyondRankEffect(5, 2, 0));
      if (hasElement(246)) start = start.add(elemEffect(246));
    } else if (name == "pe") {
      if (hasElement(233)) start = start.add(25);
    }
  } else if (type == 1) {
    if (name == "tickspeed") {
      if (player.mainUpg.rp.includes(14)) start = start.add(50);
      if (player.ranks.tetr.gte(5)) start = start.add(RANKS.effect.tetr[5]());
    } else if (name == "rank") {
      if (player.mainUpg.atom.includes(10)) start = start.add(tmp.upgs ? (tmp.upgs.main ? tmp.upgs.main[3][10].effect : 0) : 0);
    } else if (name == "prestige0") {
      if (hasElement(175)) start = start.add(30);
      if (hasElement(194)) start = start.mul(2);
    } else if (name == "hex") {
      if (hasPrestige(0, 651)) start = start.mul(4 / 3);
    }
  } else if (type == 2) {
    if (name == "rank") {
      if (hasElement(62)) start = start.add(tmp.elements.effect[62]);
    } else if (name == "tickspeed") {
      if (player.ranks.tetr.gte(5)) start = start.add(RANKS.effect.tetr[5]());
    } else if (name == "massUpg") {
      if (hasElement(189)) start = start.pow(1.5);
    } else if (name == "prestige0") {
      start = start.mul(exoticAEff(0, 1));
    }
  } else if (type == 3) {
    if (name == "rank") {
      if (player.ranks.pent.gte(1)) start = start.mul(1.1);
      if (player.ranks.pent.gte(5)) start = start.mul(RANKS.effect.pent[5]());
      if (hasPrestige(1, 5)) start = start.mul(prestigeEff(1, 5));
      start = start.mul(tmp.radiation.bs.eff[14]);
      start = start.mul(tmp.bd.upgs[4].eff);
    } else if (name == "tickspeed") {
      if (hasElement(68)) start = start.mul(2);
      if (player.ranks.pent.gte(4)) start = start.mul(RANKS.effect.pent[4]());
      start = start.mul(tmp.fermions.effs[0][5]);
      start = start.mul(getEnRewardEff(0));
      if (hasElement(158)) start = start.pow(2);
    } else if (name == "bh_condenser") {
      start = start.mul(getEnRewardEff(0));
    } else if (name == "gamma_ray") {
      start = start.mul(getEnRewardEff(0));
      if (hasPrestige(0, 867)) start = start.pow(8);
    } else if (name == "supernova") {
      if (hasPrestige(1, 2)) start = start.add(100);
    } else if (name == "tier") {
      if (hasElement(155)) start = start.mul(elemEffect(155));
      if (hasElement(181)) start = start.mul(10);
      if (hasElement(193)) start = start.mul(elemEffect(193));
    } else if (name == "tetr") {
      if (hasElement(211)) start = start.mul(elemEffect(211));
    } else if (name == "prestige0") {
      start = start.mul(exoticAEff(0, 1));
    } else if (name == "fTier") {
      if (hasAscension(0, 2)) start = start.pow(2);
    }
  } else if (type == 4) {
    if (name == "rank") {
      start = start.mul(glyphUpgEff(3));
      if (hasElement(178)) start = start.mul(elemEffect(178));
      if (hasElement(193)) start = start.mul(elemEffect(193));
    } else if (name == "supernova") {
      if (hasPrestige(0, 552)) start = start.mul(1.25);
      if (hasPrestige(3, 2)) start = start.mul(prestigeEff(3, 2));
      if (hasBeyondRank(2, 17)) start = start.mul(beyondRankEffect(2, 17)[1]);
    }
  } else if (type == 5) {
    if (name == "rank") {
      if (tmp.chal && hasBeyondRank(2, 20)) start = start.mul(tmp.chal.eff[1].scrank);
    } else if (name == "supernova") {
      if (hasBeyondRank(4, 1)) start = start.add(beyondRankEffect(4, 1, 0));
    }
  }
  if (name == "supernova" && type < 4) {
    start = start.add(tmp.prim.eff[7]);
  }
  if (name == "fTier" && type < 4) {
    if (tmp.chal && hasBeyondRank(2, 20)) start = start.mul(tmp.chal.eff[1].scrank);
  }
  if ((name == "bh_condenser" || name == "gamma_ray" || name == "tickspeed") && hasUpgrade("atom", 14)) start = start.mul(10);
  if (QCs.active() && QCM8_SCALES.includes(name)) if (!tmp.scaling_qc8.includes(name)) start = start.pow(tmp.qu.qc_eff[7][0]);
  if (hasUpgrade("br", 14) && name == "fTier" && type == 0) start = start.add(10);
  if (hasElement(88) && name == "tickspeed") start = start.mul(player.qu.rip.active ? 100 : 10);
  return start.floor();
}

function getScalingPower(type, name) {
  let power = E(1);
  if (name == "supernova" && (hasCharger(3) ? type < 5 : type < 3)) {
    power = power.mul(tmp.fermions.effs[1][4]);
  }
  if (name == "fTier" && type < 4) {
    if (hasTree("fn12")) power = power.mul(0.9);
  }
  if (name == "massUpg" && type < 2 && hasElement(84)) power = power.mul(tmp.elements.effect[84]);
  if (type == 0) {
    if (name == "rank") {
      if (player.mainUpg.rp.includes(10)) power = power.mul(0.8);
      if (player.ranks.tetr.gte(4)) power = power.mul(RANKS.effect.tetr[4]());
    } else if (name == "tier") {
      if (player.ranks.tetr.gte(4)) power = power.mul(0.8);
      if (hasElement(37)) power = power.mul(tmp.elements.effect[37]);
    } else if (name == "tetr") {
      if (hasElement(74)) power = power.mul(0.75);
    } else if (name == "massUpg") {
      if (player.mainUpg.rp.includes(8)) power = power.mul(tmp.upgs.main ? tmp.upgs.main[1][8].effect : 1);
    } else if (name == "tickspeed") {
      power = power.mul(tmp.chal ? tmp.chal.eff[1].tick : 1);
    } else if (name == "bh_condenser") {
      if (hasElement(15)) power = power.mul(0.8);
    } else if (name == "gamma_ray") {
      if (hasElement(15)) power = power.mul(0.8);
    } else if (name == "fTier") {
      if (hasTree("fn3")) power = power.mul(0.925);
    } else if (name == "cosmic_str") {
      if (hasPrestige(0, 24)) power = power.mul(0.8);
      if (hasElement(137)) power = power.mul(0.75);
      if (hasPrestige(1, 15)) power = power.mul(prestigeEff(1, 15, 1));
    } else if (name == "prestige0" || name == "prestige1") {
      if (hasElement(134)) power = power.mul(0.95);
    } else if (name == "massUpg4") {
      if (tmp.chal && hasBeyondRank(2, 20)) power = power.mul(tmp.chal.eff[1].over);
    } else if (name == "FSS") {
      if (hasBeyondRank(3, 18)) power = power.mul(beyondRankEffect(3, 18));
    }
  } else if (type == 1) {
    if (name == "rank") {
      if (player.ranks.tetr.gte(1)) power = power.mul(0.85);
      if (hasElement(27)) power = power.mul(0.75);
    } else if (name == "tier") {
      if (player.ranks.tetr.gte(4)) power = power.mul(0.8);
      if (hasElement(37)) power = power.mul(tmp.elements.effect[37]);
    } else if (name == "massUpg") {
      if (player.mainUpg.rp.includes(8)) power = power.mul(tmp.upgs.main ? tmp.upgs.main[1][8].effect : 1);
    } else if (name == "tickspeed") {
      if (player.mainUpg.bh.includes(12)) power = power.mul(0.85);
      if (hasElement(27)) power = power.mul(0.75);
    } else if (name == "bh_condenser") {
      if (hasElement(55)) power = power.mul(0.75);
    } else if (name == "gamma_ray") {
      if (hasElement(55)) power = power.mul(0.75);
    } else if (name == "cosmic_str") {
      if (hasElement(137)) power = power.mul(0.75);
      if (hasPrestige(1, 15)) power = power.mul(prestigeEff(1, 15, 1));
    } else if (name == "prestige0") {
      if (hasElement(154)) power = power.mul(0.9);
    } else if (name == "tetr") {
      if (hasElement(154)) power = power.mul(0.9);
    } else if (name == "pent") {
      if (hasElement(154)) power = power.mul(0.9);
    }
  } else if (type == 2) {
    if (name == "rank") {
      if (hasElement(27)) power = power.mul(0.75);
      if (hasElement(58)) power = power.mul(tmp.elements.effect[58]);
    } else if (name == "tickspeed") {
      if (hasElement(27)) power = power.mul(0.75);
      if (hasElement(58)) power = power.mul(tmp.elements.effect[58]);
    } else if (name == "bh_condenser") {
      if (hasElement(55)) power = power.mul(0.75);
    } else if (name == "gamma_ray") {
      if (hasElement(55)) power = power.mul(0.75);
    } else if (name == "prestige0") {
      if (hasElement(197)) power = power.mul(0.9);
      if (tmp.chal && hasCharger(3)) power = power.mul(tmp.chal.eff[5]);
    }
  } else if (type == 3) {
    if (name == "supernova") {
      if (hasElement(78)) power = power.mul(0.8);
    }
  } else if (type == 4) {
    if (name == "rank") {
      if (hasElement(197)) power = power.mul(0.9);
      if (tmp.chal && hasCharger(3)) power = power.mul(tmp.chal.eff[5]);
    } else if (name == "tier") {
      if (tmp.chal && hasCharger(3)) power = power.mul(tmp.chal.eff[5]);
    } else if (name == "supernova") {
      if (hasElement(212)) power = power.mul(0.75);
    }
  }
  if (hasUpgrade("atom", 15) && name == "gamma_ray") power = power.mul(0.8);
  if (hasElement(108) && ["rank", "tier", "tetr", "pent"].includes(name) && type < 4) power = power.mul(player.qu.rip.active ? 0.98 : 0.9);

  let rps = ["rank", "tier", "tetr", "pent"];
  if (hasElement(207)) rps.push("hex");
  if (hasPrestige(2, 4) && rps.includes(name) && (name == "hex" ? type < 2 : type < 4)) power = power.mul(tmp.qu.chroma_eff[1][1]);

  let qf = tmp.qu.qc_eff[7][1];
  if (!tmp.c16active) if (player.dark.run.upg[4] && inDarkRun() && ["rank", "tier", "tetr", "pent", "hex"].includes(name)) qf **= 0.75;
  if (QCs.active() && QCM8_SCALES.includes(name)) if (!tmp.scaling_qc8.includes(name)) power = power.mul(qf);
  if (PreQ_SCALES.includes(name) && type < 3) power = power.mul(getEnRewardEff(5));

  let p = ["prestige0", "prestige1"];
  if (hasPrestige(3, 10)) p.push("prestige2");
  if (hasPrestige(0, 388) && p.includes(name) && type < 3) power = power.mul(prestigeEff(0, 388, 1));

  if (hasPrestige(1, 66) && name == "fTier") power = power.mul(0.8);
  return power.max(type == 3 ? 0.5 : 0);
}

function noScalings(type, name) {
  if (name == "rank") {
    if (type < 4 && hasPrestige(1, 127)) return true;
    // else if (type == 4) return true
  } else if (name == "tier") {
    if (type < 4 && hasPrestige(1, 127)) return true;
  } else if (name == "tetr") {
    return hasCharger(8);
  } else if (name == "pent") {
    return hasElement(243);
  } else if (name == "massUpg") {
    if (hasBeyondRank(2, 15)) return true;
  } else if (name == "supernova") {
    if (type < 3 && hasCharger(3)) return true;
  } else if (name == "tickspeed") {
    if (hasCharger(4)) return true;
  } else if (name == "fTier") {
    if (type < 3 && hasElement(2, 1)) return true;
  } else if (name == "bh_condenser") {
    if (hasCharger(6)) return true;
  } else if (name == "gamma_ray") {
    if (hasCharger(7)) return true;
  } else if (name == "prestige0") {
    if (type < 3 && hasBeyondRank(5, 7)) return true;
  }

  return false;
}
